lib/Extractor/Candidates.cpp

```diff
@@ -767,12 +767,14 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
   for (auto &BB : F) {
     std::unique_ptr<BlockCandidateSet> BCS(new BlockCandidateSet);
     for (auto &I : BB) {
-      if (I.getType()->isIntegerTy()) {
-        Inst *In = EB.get(&I);
-        EB.markExternalUses(In);
-        BCS->Replacements.emplace_back(&I, InstMapping(In, 0));
-        assert(EB.get(&I)->hasOrigin(&I));
-      }
+      if (!I.getType()->isIntegerTy())
+        continue;
+      if (I.hasNUses(0))
```

Comment:
What is the rationale for this?  
It is possible that a new RHS could utilize a value which was not used.  
That is easier than synthesizing it from scratch, especially if constants are involved.

Reply:
this patch causes souper to refuse to optimize a value that isn't used anywhere

you're talking about a different issue, which is where some value that is used could perhaps be computed in a way that takes advantage of a value that isn't used, but which happens to be sitting around. making souper use those is a different problem.

and anyway, LLVM aggressively cleans up unused values when possible, so souper seeing them would be a matter of chance and phase ordering.

Reply:
Okay, maybe it would make sense to harvest a set of 'components' independently from the LHS candidate.   
LGTM

Reply:
this might work-- but let's have the conversation elsewhere